src/libsyntax/parse/mod.rs

```diff
@@ -318,7 +310,10 @@ pub fn char_lit(lit: &str) -> (char, isize) {
 
     let unicode_escape = || -> Option<(char, isize)> {
         if lit.as_bytes()[2] == b'{' {
-            let idx = lit.find('}').expect(msg2);
+            let idx = match lit.find('}') {
+                Some(idx2) => idx2,
+                None => panic!("lexer should have rejected a bad character escape {}", lit),
+            };
```

Comment:
You could make this a bit less verbose while still keeping the speed improvement with `unwrap_or_else`:
```rust
let idx = lit.find('}').unwrap_or_else(|| {
    panic!("lexer should have rejected a bad character escape {}", lit)
});
```